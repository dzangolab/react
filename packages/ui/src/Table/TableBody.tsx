import { Cell, flexRender, Table, NoInfer } from "@tanstack/react-table";
import React from "react";

import {
  TableCell,
  TableRow,
  TooltipWrapper,
  TableBody as TTableBody,
} from "./TableElements";
import { formatDate, formatNumber, getAlignValue } from "./utils";

import type { TDataTableProperties } from "./types";

interface THeaderProperty<T>
  extends Pick<
    TDataTableProperties<T>,
    "enableRowSelection" | "customFormatters" | "emptyTableMessage"
  > {
  parsedColumnsLength: number;
  table: Table<T>;
}

export const TableBody = <TData extends { id: string | number }>({
  customFormatters = {},
  emptyTableMessage = "No results.",
  enableRowSelection,
  parsedColumnsLength,
  table,
}: THeaderProperty<TData>) => {
  const renderTooltipContent = (
    cell: Cell<TData, unknown>,
  ): React.ReactNode => {
    const tooltip = cell.column.columnDef.tooltip;

    if (typeof tooltip === "string") {
      return tooltip;
    } else if (typeof tooltip === "function") {
      return tooltip(cell);
    }

    return cell.getValue() as string;
  };

  return (
    <TTableBody>
      {table.getRowModel().rows?.length ? (
        table.getRowModel().rows.map((row) => (
          <TableRow
            key={row.id}
            {...(enableRowSelection && {
              "data-selected": row.getIsSelected(),
            })}
            data-id={row.original.id ?? row.id}
          >
            {row.getVisibleCells().map((cell) => {
              const getFormattedValueContext: typeof cell.getContext = () => {
                const cellContext = cell.getContext();
                const renderValue = cellContext.getValue;
                const dateOptions = cell.column.columnDef.dateOptions;
                const numberOptions = cell.column.columnDef.numberOptions;

                const getFormattedValue = (): NoInfer<never> => {
                  const defaultCustomFormatters: Record<
                    string,
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    (value: any) => NoInfer<never>
                  > = {
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    number: (value: any) =>
                      formatNumber({
                        value: Number(value),
                        locale: numberOptions?.locale,
                        formatOptions: numberOptions?.formatOptions,
                      }) as NoInfer<never>,
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    date: (value: any) =>
                      formatDate({
                        date: value as Date,
                        locale: dateOptions?.locale,
                        formatOptions: dateOptions?.formatOptions,
                      }) as NoInfer<never>,
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    currency: (value: any) =>
                      formatNumber({
                        value: Number(value),
                        locale: numberOptions?.locale,
                        formatOptions: {
                          style: "currency",
                          currency: "USD",
                          ...(numberOptions?.formatOptions &&
                            numberOptions.formatOptions),
                        },
                      }) as NoInfer<never>,
                    ...customFormatters,
                  };

                  const dataType: string =
                    cell.column.columnDef.dataType || "text";

                  return (
                    defaultCustomFormatters?.[dataType]?.(renderValue()) ||
                    renderValue()
                  );
                };

                return {
                  ...cellContext,
                  renderValue: () => getFormattedValue(),
                  getValue: () => getFormattedValue(),
                };
              };

              return (
                <TableCell
                  key={cell.id}
                  data-label={cell.column.id}
                  data-align={getAlignValue({
                    align: cell.column.columnDef.align,
                    dataType: cell.column.columnDef.dataType,
                  })}
                  className={`${
                    cell.column.id ? `cell-${cell.column.id}` : ``
                  } ${cell.column.columnDef.className || ""}`
                    .replace(/\s\s/, " ")
                    .trimEnd()}
                  style={{
                    width: cell.column.columnDef.width,
                    maxWidth: cell.column.columnDef.maxWidth,
                    minWidth: cell.column.columnDef.minWidth,
                  }}
                >
                  {cell.column.columnDef.tooltip ? (
                    <TooltipWrapper
                      tooltipOptions={{
                        children: renderTooltipContent(cell),
                        ...cell.column.columnDef?.tooltipOptions,
                      }}
                      cellContent={flexRender(
                        cell.column.columnDef.cell,
                        getFormattedValueContext(),
                      )}
                    ></TooltipWrapper>
                  ) : (
                    flexRender(
                      cell.column.columnDef.cell,
                      getFormattedValueContext(),
                    )
                  )}
                </TableCell>
              );
            })}
          </TableRow>
        ))
      ) : (
        <TableRow>
          <TableCell colSpan={parsedColumnsLength}>
            {emptyTableMessage}
          </TableCell>
        </TableRow>
      )}
    </TTableBody>
  );
};
