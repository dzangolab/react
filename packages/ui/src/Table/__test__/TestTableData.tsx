const data = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
];

const biggerData = [
  { id: 1, name: "jone" },
  { id: 2, name: "abhi" },
  { id: 3, name: "mike" },
  { id: 4, name: "tony" },
  { id: 5, name: "tony" },
  { id: 6, name: "tony" },
  { id: 7, name: "tony" },
  { id: 8, name: "tony" },
  { id: 9, name: "tony" },
  { id: 10, name: "tony" },
  { id: 11, name: "tony" },
  { id: 12, name: "tony" },
  { id: 13, name: "tony" },
  { id: 14, name: "tony" },
  { id: 15, name: "tony" },
];

const fetcher = () => {
  return data;
};

const columns = [
  {
    accessorKey: "id",
    header: () => <span>ID</span>,
  },
  {
    accessorKey: "name",
    header: () => <span>Name</span>,
  },
];

export { columns, data, fetcher, biggerData };
