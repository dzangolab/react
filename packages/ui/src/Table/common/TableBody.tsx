import { flexRender } from "@tanstack/react-table";
import React, { useContext } from "react";

import { TableContext } from "./TableContext";

function TableBody() {
  const { table } = useContext(TableContext);

  return (
    <tbody>
      {table?.getRowModel().rows.map((row) => {
        const id = row.original.id ? row.original.id : row.id;

        return (
          <tr key={id} data-id={id}>
            {row.getVisibleCells().map((cell) => {
              return (
                <td key={cell.id} data-label={cell.column.id}>
                  {flexRender(cell.column.columnDef.cell, cell.getContext())}
                </td>
              );
            })}
          </tr>
        );
      })}
    </tbody>
  );
}

export default TableBody;
