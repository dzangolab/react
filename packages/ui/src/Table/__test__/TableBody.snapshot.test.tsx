import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test } from "vitest";

import { columns, data, fetcher } from "./TestTableData";
import TableBody from "../common/TableBody";
import { Table } from "../index";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <Table
      columns={columns}
      data={data}
      fetcher={fetcher}
      totalItems={data.length}
    >
      <TableBody />
    </Table>
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
