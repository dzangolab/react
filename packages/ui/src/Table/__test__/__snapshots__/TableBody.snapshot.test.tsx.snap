// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<tbody>
  <tr>
    <td
      data-label="id"
    >
      1
    </td>
    <td
      data-label="name"
    >
      jone
    </td>
  </tr>
  <tr>
    <td
      data-label="id"
    >
      2
    </td>
    <td
      data-label="name"
    >
      abhi
    </td>
  </tr>
  <tr>
    <td
      data-label="id"
    >
      3
    </td>
    <td
      data-label="name"
    >
      mike
    </td>
  </tr>
  <tr>
    <td
      data-label="id"
    >
      4
    </td>
    <td
      data-label="name"
    >
      tony
    </td>
  </tr>
</tbody>
`;
